security/user-last-seen.php

```diff
@@ -0,0 +1,330 @@
+		}
+
+		return $user;
+	}
+
+	public function add_last_seen_column_head( $columns ) {
+		$columns['last_seen'] = __( 'Last seen', 'wpvip' );
+		return $columns;
+	}
+
+	public function add_last_seen_sortable_column( $columns ) {
+		$columns['last_seen'] = 'last_seen';
+
+		return $columns;
+	}
+
+	public function last_seen_order_by_query_args( $vars ) {
+		if ( isset( $vars['orderby'] ) && 'last_seen' === $vars['orderby'] ) {
+			$vars['meta_key'] = self::LAST_SEEN_META_KEY;
+			$vars['orderby']  = 'meta_value_num';
+		}
+
+		return $vars;
+	}
+
+	public function last_seen_blocked_users_filter_query_args( $vars ) {
+		if ( isset( $_GET['last_seen_filter'] ) && 'blocked' === $_GET['last_seen_filter'] && isset( $_GET['last_seen_filter_nonce'] ) && wp_verify_nonce( sanitize_text_field( $_GET['last_seen_filter_nonce'] ), 'last_seen_filter' ) ) {
+			$vars['meta_key'] = self::LAST_SEEN_META_KEY;
+			// phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_value
+			$vars['meta_value']   = $this->get_inactivity_timestamp();
+			$vars['meta_type']    = 'NUMERIC';
+			$vars['meta_compare'] = '<';
+		}
+
+		return $vars;
+	}
+
+	public function add_last_seen_column_date( $default, $column_name, $user_id ) {
+		if ( 'last_seen' !== $column_name ) {
+			return $default;
+		}
+
+		$last_seen_timestamp = get_user_meta( $user_id, self::LAST_SEEN_META_KEY, true );
+
+		if ( ! $last_seen_timestamp ) {
+			return $default;
+		}
+
+		$date = sprintf(
+			/* translators: 1: Comment date, 2: Comment time. */
+			__( '%1$s at %2$s' ),
+			date_i18n( get_option( 'date_format' ), $last_seen_timestamp ),
+			date_i18n( get_option( 'time_format' ), $last_seen_timestamp )
+		);
+
+		if ( ! $this->is_block_action_enabled() || ! $this->is_considered_inactive( $user_id ) ) {
+			return sprintf( '<span>%s</span>', esc_html( $date ) );
+		}
+
+		$unblock_link = '';
+		if ( current_user_can( 'edit_user', $user_id ) ) {
+			$url = add_query_arg( array(
+				'action'                => 'reset_last_seen',
+				'user_id'               => $user_id,
+				'reset_last_seen_nonce' => wp_create_nonce( 'reset_last_seen_action' ),
+			) );
+
+			$unblock_link = "<div class='row-actions'><span>User blocked due to inactivity. <a class='reset_last_seen_action' href='" . esc_url( $url ) . "'>" . __( 'Unblock', 'wpvip' ) . '</a></span></div>';
+		}
+		return sprintf( '<span class="wp-ui-text-notification">%s</span>' . $unblock_link, esc_html( $date ) );
+	}
+
+	public function add_blocked_users_filter( $views ) {
+		$blog_id = is_network_admin() ? null : get_current_blog_id();
+
+		$users_query = new \WP_User_Query(
+			array(
+				'blog_id'      => $blog_id,
+				'fields'       => 'ID',
+				'meta_key'     => self::LAST_SEEN_META_KEY,
+				// phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_value
+				'meta_value'   => $this->get_inactivity_timestamp(),
```

Comment:
Because this number always changes, it's not able to take advantage of any query caching 😬 . TBD on performance for this on larger sites, but might want to re-visit layer and add a simple caching layer around it.